lib/rules/no-constant-binary-operand.js

```diff
@@ -0,0 +1,431 @@
+                node.operator === "??="
+            ) {
+                return false;
+            }
+
+            /*
+             * The remaining assignment expressions all result in a numeric or
+             * string (non-nullish) value:
+             *   "+=", "-=", "*=", "/=", "%=", "<<=", ">>=", ">>>=", "|=", "^=", "&="
+             */
+
+            return true;
+        case "UnaryExpression":
+
+            /*
+             * "void" Always returns `undefined`
+             * "typeof" All types are strings, and thus non-nullish
+             * "!" Boolean is never nullish
+             * "delete" Returns a boolean, which is never nullish
+             * Math operators always return numbers or strings, neither of which
+             * are non-nullish "+", "-", "~"
+             */
+
+            return true;
+        case "SequenceExpression": {
+            const last = node.expressions[node.expressions.length - 1];
+
+            return hasConstantNullishness(last);
+        }
+        case "Identifier": {
+            return node.name === "undefined";
+        }
+        case "JSXElement": // ESLint has a policy of not assuming any specific JSX behavior.
+        case "JSXFragment":
+            return false;
+        default:
+            return false;
+    }
+}
+
+/**
+ * Test if an AST node is a boolean value that never changes
+ * @param {ASTNode} node The node to test
+ * @returns {boolean} Is `node` guaranteed to be a boolean?
+ */
+function isStaticBoolean(node) {
+    switch (node.type) {
+        case "Literal":
+            return typeof node.value === "boolean";
+        case "CallExpression":
+            return node.callee.name === "Boolean" &&
+         (node.arguments.length === 0 || hasConstantTruthiness(node.arguments[0]));
+        case "UnaryExpression":
+            return node.operator === "!" && hasConstantTruthiness(node.argument);
+        default:
+            return false;
+    }
+}
+
+
+/**
+ * Test if an AST node will always give the same result when compared to a
+ * bolean value. Note that comparison to boolean values is different to
+ * truthiness.
+ * https://262.ecma-international.org/5.1/#sec-11.9.3
+ *
+ * Javascript `==` operator works by converting the boolean to `1` (true) or
+ * `+0` (false) and then checks the values `==` equality to that number.
+ * @param {ASTNode} node The node to test
+ * @returns {boolean} Will `node` always coerce to the same boolean value?
+ */
+function hasConstantLooseBooleanComparison(node) {
+    switch (node.type) {
+        case "ObjectExpression":
+            return !node.properties.some(prop => OBJECT_EQUALITY_KEYS.has(prop.key.name));
+        case "ArrayExpression":
+            if (node.elements.length === 1) {
+
+                /*
+                 * TODO: We could check if the single value would result in
+                 * variable boolean comparison.
+                 */
+
+                return false; // `[0]` and `[1]` would compare differently.
```

Comment:
I'm not sure I understand this.
Won't `[0]` be converted to `'0'` (and similarly for `[1]`) - which are constant strings.
Am I misunderstanding the conversion?

Reply:
Yeah, an array with statically declared elements would be fine (that's what the TODO is trying to call out), but `[x]` cannot be reasoned about statically since:

```
[1] == true
=> true
[0] == true
=> false
```

Reply:
I've updated the comment to be a little more clear.